File: scripts/base_controller_node.py
#!/usr/bin/env python3
import os
import sys
import rospy
import math
import time
import tf
from nav_msgs.msg import Odometry
from geometry_msgs.msg import Quaternion, Twist
from std_msgs.msg import Float32MultiArray

# Adjust the path so Python can locate the ddsm115 module.
this_dir = os.path.dirname(os.path.abspath(__file__))
ddsm115_path = os.path.abspath(os.path.join(this_dir, '..', 'ddsm115'))
if ddsm115_path not in sys.path:
    sys.path.insert(0, ddsm115_path)

import ddsm115

class BaseController:
    def __init__(self):
        rospy.init_node("base_controller", anonymous=True)

        # Load parameters
        self.wheel_names = rospy.get_param("~wheel_names", 
            ["front_left_wheel", "front_right_wheel", "rear_left_wheel", "rear_right_wheel"])
        self.wheel_ids = rospy.get_param("~wheel_ids", [1, 2, 3, 4])
        self.wheel_directions = rospy.get_param("~wheel_directions", 
            ["forward", "backward", "forward", "backward"])
        self.wheel_radius = rospy.get_param("~wheel_radius", 0.1)      # in meters
        self.wheel_track = rospy.get_param("~wheel_track", 0.5)        # meters (distance between left & right wheels)
        self.wheel_base = rospy.get_param("~wheel_base", 0.6)          # meters (distance between front & rear wheels)
        self.feedback_rate = rospy.get_param("~feedback_rate", 100)    # Hz

        # New parameters for slight turning behavior.
        self.slight_turn_threshold = rospy.get_param("~slight_turn_threshold", 0.1)  # rad/s threshold for slight turn
        self.inside_reduction_factor = rospy.get_param("~inside_reduction_factor", 0.3)  # fraction to reduce inside wheel speed

        # Instantiate the motor control object from ddsm115.
        self.motor_control = ddsm115.MotorControl()

        # Set all motors to drive mode 2.
        for motor_id in self.wheel_ids:
            self.motor_control.set_drive_mode(motor_id, 2)

        # Initialize odometry state.
        self.x = 0.0
        self.y = 0.0
        self.theta = 0.0
        self.last_time = rospy.Time.now()

        # Publishers for odometry, motor RPMs, and motor currents, and TF broadcaster.
        self.odom_pub = rospy.Publisher("wheel_odom", Odometry, queue_size=10)
        self.odom_broadcaster = tf.TransformBroadcaster()
        self.motor_rpm_pub = rospy.Publisher("motor_rpms", Float32MultiArray, queue_size=10)
        self.motor_current_pub = rospy.Publisher("motor_currents", Float32MultiArray, queue_size=10)

        # Subscriber for /cmd_vel commands.
        self.cmd_vel_sub = rospy.Subscriber("cmd_vel", Twist, self.cmd_vel_callback)

        # Timer for the feedback loop.
        self.timer = rospy.Timer(rospy.Duration(1.0 / self.feedback_rate), self.update)

    def cmd_vel_callback(self, msg):
        """
        Converts incoming /cmd_vel commands into motor RPM commands.

        For slight turns (angular velocity below slight_turn_threshold), instead of applying
        full differential drive offsets, the inside wheels are commanded at a reduced speed
        relative to the current forward speed.

        For larger rotations, the standard differential drive kinematics are used:
            left_vel  = v - (w * wheel_track/2)
            right_vel = v + (w * wheel_track/2)
        
        Linear velocities (m/s) are then converted to RPM.
        """
        v = msg.linear.x
        w = msg.angular.z

        conversion_factor = 60.0 / (2 * math.pi * self.wheel_radius)

        # Check if this is a slight turn.
        if abs(w) < self.slight_turn_threshold:
            # For a slight turn, reduce the speed of the inside wheels.
            if w > 0:
                # Turning left: left wheels are inside.
                left_vel = v * (1 - self.inside_reduction_factor)
                right_vel = v
            elif w < 0:
                # Turning right: right wheels are inside.
                left_vel = v
                right_vel = v * (1 - self.inside_reduction_factor)
            else:
                left_vel = v
                right_vel = v
        else:
            # Standard differential drive for larger rotations.
            left_vel = v - (w * self.wheel_track / 2.0)
            right_vel = v + (w * self.wheel_track / 2.0)

        # Convert linear velocities to RPM.
        left_rpm = left_vel * conversion_factor
        right_rpm = right_vel * conversion_factor

        # Command each motor with the appropriate RPM.
        for i, motor_id in enumerate(self.wheel_ids):
            rpm = left_rpm if i in [0, 2] else right_rpm

            # Adjust for wheel direction.
            if self.wheel_directions[i] == "backward":
                rpm = -rpm

            try:
                self.motor_control.send_rpm(motor_id, rpm)
            except Exception as e:
                rospy.logwarn("Error sending rpm to motor {}: {}".format(motor_id, e))

    def update(self, event):
        current_time = rospy.Time.now()
        dt = (current_time - self.last_time).to_sec()
        self.last_time = current_time

        wheel_velocities = []
        motor_rpms = []
        motor_currents = []

        # Read feedback for each motor.
        for i, motor_id in enumerate(self.wheel_ids):
            try:
                feedback = self.motor_control.get_motor_feedback(motor_id)
                rpm = feedback[0]
                current = feedback[1]
            except Exception as e:
                rospy.logwarn("Error reading feedback for motor {}: {}".format(motor_id, e))
                rpm, current = 0, 0

            # Adjust RPM based on wheel direction.
            if self.wheel_directions[i] == "backward":
                rpm = -rpm

            motor_rpms.append(rpm)
            motor_currents.append(current)
            # Convert rpm to linear velocity (m/s): v = rpm * (2*pi*radius) / 60.
            v_conv = rpm * (2 * math.pi * self.wheel_radius) / 60.0
            wheel_velocities.append(v_conv)

        # For odometry, assume wheels 0 and 2 are on the left, and wheels 1 and 3 are on the right.
        v_left = (wheel_velocities[0] + wheel_velocities[2]) / 2.0
        v_right = (wheel_velocities[1] + wheel_velocities[3]) / 2.0

        linear_velocity = (v_left + v_right) / 2.0
        angular_velocity = (v_right - v_left) / self.wheel_track

        # Update the robot's pose (simple Euler integration).
        self.x += linear_velocity * math.cos(self.theta) * dt
        self.y += linear_velocity * math.sin(self.theta) * dt
        self.theta += angular_velocity * dt

        # Publish the odometry message.
        odom = Odometry()
        odom.header.stamp = current_time
        odom.header.frame_id = "odom"
        odom.pose.pose.position.x = self.x
        odom.pose.pose.position.y = self.y
        odom.pose.pose.position.z = 0.0
        q = tf.transformations.quaternion_from_euler(0, 0, self.theta)
        odom.pose.pose.orientation = Quaternion(*q)
        odom.child_frame_id = "base_link"
        odom.twist.twist.linear.x = linear_velocity
        odom.twist.twist.angular.z = angular_velocity
        self.odom_pub.publish(odom)
        self.odom_broadcaster.sendTransform(
            (self.x, self.y, 0),
            q,
            current_time,
            "base_link",
            "odom"
        )

        # Publish motor RPM and current topics.
        rpm_msg = Float32MultiArray()
        rpm_msg.data = motor_rpms
        self.motor_rpm_pub.publish(rpm_msg)

        current_msg = Float32MultiArray()
        current_msg.data = motor_currents
        self.motor_current_pub.publish(current_msg)

if __name__ == '__main__':
    try:
        controller = BaseController()
        rospy.spin()
    except rospy.ROSInterruptException:
        pass

File: src/base_controller_node.cpp
/**
 * base_controller_node.cpp
 *
 * C++ ROS node for DDSM115 4WD base controller.
 *
 * Architecture:
 *  - cmd_vel callback:  stores velocity command, signals TX thread
 *  - TX thread:         converts cmd to RPM, sends to all 4 motors with
 *                       inter-command gap to prevent RS485 bus collisions
 *  - Feedback thread:   polls motor feedback at fixed rate, publishes odometry
 *  - bus_mutex_:        ensures TX and feedback never overlap on the serial bus
 */

#include <ros/ros.h>
#include <geometry_msgs/Twist.h>
#include <nav_msgs/Odometry.h>
#include <std_msgs/Float32MultiArray.h>
#include <tf/transform_broadcaster.h>

#include <serial/serial.h>

#include <atomic>
#include <chrono>
#include <condition_variable>
#include <cmath>
#include <mutex>
#include <string>
#include <thread>
#include <vector>

#include "robot_4wd_dds115_base_controller/ddsm115_protocol.hpp"

class BaseControllerNode
{
public:
    BaseControllerNode() : pnh_("~"), running_(false)
    {
        // ── Parameters ───────────────────────────────────────────────────────
        std::string device;
        pnh_.param<std::string>("device_path",    device,            "/dev/rs485");
        pnh_.param<double>("wheel_radius",        wheel_radius_,     0.05);
        pnh_.param<double>("wheel_track",         wheel_track_,      0.33);
        pnh_.param<double>("wheel_base",          wheel_base_,       0.33);
        pnh_.param<double>("feedback_rate_hz",    feedback_rate_hz_, 10.0);
        pnh_.param<double>("cmd_vel_timeout",     cmd_vel_timeout_,  1.0);
        pnh_.param<int>   ("inter_cmd_gap_ms",    inter_cmd_gap_ms_, 2);

        pnh_.param("wheel_ids",        wheel_ids_,        std::vector<int>{1, 2, 3, 4});
        pnh_.param("wheel_directions", wheel_directions_,
                   std::vector<std::string>{"forward", "backward", "forward", "backward"});

        // ── Serial port ───────────────────────────────────────────────────────
        try {
            serial_.setPort(device);
            serial_.setBaudrate(115200);
            serial::Timeout to = serial::Timeout::simpleTimeout(50);  // 50ms read timeout
            serial_.setTimeout(to);
            serial_.open();
        } catch (const serial::IOException& e) {
            ROS_FATAL("Failed to open serial port %s: %s", device.c_str(), e.what());
            ros::shutdown();
            return;
        }
        ROS_INFO("Opened serial port: %s", device.c_str());

        // ── Set all motors to velocity mode ──────────────────────────────────
        for (int id : wheel_ids_) {
            auto frame = ddsm115::buildDriveModeFrame(
                static_cast<uint8_t>(id), ddsm115::DriveMode::VELOCITY);
            serial_.write(frame.data(), frame.size());
            std::this_thread::sleep_for(std::chrono::milliseconds(20));
            serial_.flushInput();
            ROS_INFO("Motor %d set to velocity mode", id);
        }

        // ── ROS interfaces ────────────────────────────────────────────────────
        cmd_vel_sub_  = nh_.subscribe("cmd_vel",      10,
                                      &BaseControllerNode::cmdVelCallback, this);
        odom_pub_     = nh_.advertise<nav_msgs::Odometry>("wheel_odom", 10);
        rpm_pub_      = nh_.advertise<std_msgs::Float32MultiArray>("motor_rpms", 10);
        current_pub_  = nh_.advertise<std_msgs::Float32MultiArray>("motor_currents", 10);

        // ── Init state ────────────────────────────────────────────────────────
        motor_rpms_.fill(0);
        motor_currents_.fill(0.0f);
        last_cmd_time_ = ros::Time::now();
        last_odom_time_ = ros::Time::now();

        // ── Start worker threads ──────────────────────────────────────────────
        running_ = true;
        tx_thread_       = std::thread(&BaseControllerNode::txThreadFunc, this);
        feedback_thread_ = std::thread(&BaseControllerNode::feedbackThreadFunc, this);

        ROS_INFO("Base controller ready.");
    }

    ~BaseControllerNode()
    {
        running_ = false;
        cmd_cv_.notify_all();
        if (tx_thread_.joinable())       tx_thread_.join();
        if (feedback_thread_.joinable()) feedback_thread_.join();

        // Stop all motors before exit
        if (serial_.isOpen()) {
            std::lock_guard<std::mutex> lock(bus_mutex_);
            for (int id : wheel_ids_) {
                auto frame = ddsm115::buildRpmFrame(static_cast<uint8_t>(id), 0);
                serial_.write(frame.data(), frame.size());
                std::this_thread::sleep_for(
                    std::chrono::milliseconds(inter_cmd_gap_ms_));
                serial_.flushInput();
            }
            serial_.close();
        }
    }

    void spin() { ros::spin(); }

private:
    // ── cmd_vel callback ─────────────────────────────────────────────────────
    void cmdVelCallback(const geometry_msgs::Twist::ConstPtr& msg)
    {
        {
            std::lock_guard<std::mutex> lock(cmd_mutex_);
            cmd_linear_  = msg->linear.x;
            cmd_angular_ = msg->angular.z;
            cmd_updated_ = true;
            last_cmd_time_ = ros::Time::now();
        }
        cmd_cv_.notify_one();
    }

    // ── TX thread ─────────────────────────────────────────────────────────────
    // Wakes on new cmd_vel or every 50ms (watchdog keepalive).
    void txThreadFunc()
    {
        while (running_) {
            double linear, angular;
            bool   watchdog_stop = false;

            {
                std::unique_lock<std::mutex> lock(cmd_mutex_);
                cmd_cv_.wait_for(lock, std::chrono::milliseconds(50));

                if (!running_) break;

                linear  = cmd_linear_;
                angular = cmd_angular_;
                cmd_updated_ = false;

                // Watchdog: if no cmd received recently, send zero
                if (cmd_vel_timeout_ > 0.0) {
                    double elapsed = (ros::Time::now() - last_cmd_time_).toSec();
                    if (elapsed > cmd_vel_timeout_) {
                        linear = angular = 0.0;
                        watchdog_stop = true;
                    }
                }
            }

            sendRpms(linear, angular);

            if (watchdog_stop) {
                ROS_WARN_THROTTLE(5.0, "cmd_vel watchdog: stopping motors");
            }
        }
    }

    // ── Send RPMs to all 4 motors ─────────────────────────────────────────────
    // Pure differential drive kinematics. No threshold hacks.
    void sendRpms(double linear, double angular)
    {
        const double conv = 60.0 / (2.0 * M_PI * wheel_radius_);
        const double left_vel  = linear - angular * wheel_track_ / 2.0;
        const double right_vel = linear + angular * wheel_track_ / 2.0;

        std::lock_guard<std::mutex> bus_lock(bus_mutex_);

        for (size_t i = 0; i < wheel_ids_.size(); i++) {
            double vel = (wheel_directions_[i] == "backward") ? -right_vel : left_vel;
            if (i == 1 || i == 3) {
                // right-side motors
                vel = (wheel_directions_[i] == "backward") ? -right_vel : right_vel;
            }
            auto rpm_val = static_cast<int16_t>(std::round(vel * conv));
            auto frame   = ddsm115::buildRpmFrame(
                               static_cast<uint8_t>(wheel_ids_[i]), rpm_val);

            serial_.flushInput();
            serial_.write(frame.data(), frame.size());

            // Wait for RS485 line to clear before next command
            std::this_thread::sleep_for(
                std::chrono::milliseconds(inter_cmd_gap_ms_));
        }
        // Flush any accumulated replies — we don't need them here
        serial_.flushInput();
    }

    // ── Feedback thread ───────────────────────────────────────────────────────
    void feedbackThreadFunc()
    {
        const auto period = std::chrono::duration<double>(1.0 / feedback_rate_hz_);
        auto next_wake    = std::chrono::steady_clock::now() + period;

        while (running_) {
            {
                std::lock_guard<std::mutex> bus_lock(bus_mutex_);
                for (size_t i = 0; i < wheel_ids_.size(); i++) {
                    pollFeedback(static_cast<uint8_t>(wheel_ids_[i]),
                                 static_cast<int>(i));
                }
            }
            publishOdometry();
            publishMotorFeedback();

            std::this_thread::sleep_until(next_wake);
            next_wake += period;
        }
    }

    // ── Poll one motor for feedback ───────────────────────────────────────────
    void pollFeedback(uint8_t motor_id, int idx)
    {
        auto req = ddsm115::buildFeedbackRequestFrame(motor_id);
        serial_.flushInput();
        serial_.write(req.data(), req.size());

        // Read reply: scan for a valid 10-byte frame starting with motor_id
        ddsm115::Frame reply{};
        size_t filled = 0;
        auto deadline = std::chrono::steady_clock::now() +
                        std::chrono::milliseconds(50);

        while (filled < ddsm115::FRAME_SIZE &&
               std::chrono::steady_clock::now() < deadline)
        {
            uint8_t byte;
            size_t  n = serial_.read(&byte, 1);
            if (n == 0) continue;

            if (filled == 0 && byte != motor_id) continue;  // wait for ID
            reply[filled++] = byte;
        }

        if (filled < ddsm115::FRAME_SIZE) {
            ROS_WARN_THROTTLE(2.0, "Motor %d: feedback timeout", motor_id);
            return;
        }

        auto fb = ddsm115::parseFeedback(reply, motor_id);
        if (!fb.valid) {
            ROS_WARN_THROTTLE(2.0, "Motor %d: invalid feedback (CRC or mode mismatch)",
                              motor_id);
            return;
        }

        if (fb.error) {
            ROS_WARN_THROTTLE(1.0,
                "Motor %d error flags: 0x%02X "
                "(sensor=%d overcurrent=%d phase=%d stall=%d)",
                motor_id, fb.error,
                fb.error & 0x01, (fb.error >> 1) & 0x01,
                (fb.error >> 2) & 0x01, (fb.error >> 3) & 0x01);
        }

        // Apply direction convention
        int16_t rpm = fb.rpm;
        if (wheel_directions_[static_cast<size_t>(idx)] == "backward") rpm = -rpm;

        {
            std::lock_guard<std::mutex> lock(feedback_mutex_);
            motor_rpms_[idx]     = rpm;
            motor_currents_[idx] = fb.current;
        }
    }

    // ── Publish odometry ─────────────────────────────────────────────────────
    void publishOdometry()
    {
        std::array<int16_t, 4> rpms;
        {
            std::lock_guard<std::mutex> lock(feedback_mutex_);
            rpms = motor_rpms_;
        }

        ros::Time now = ros::Time::now();
        double dt = (now - last_odom_time_).toSec();
        last_odom_time_ = now;

        if (dt <= 0.0 || dt > 1.0) return;

        const double conv  = (2.0 * M_PI * wheel_radius_) / 60.0;
        double v_left  = (rpms[0] + rpms[2]) / 2.0 * conv;
        double v_right = (rpms[1] + rpms[3]) / 2.0 * conv;

        double linear  = (v_left + v_right) / 2.0;
        double angular = (v_right - v_left) / wheel_track_;

        x_     += linear * std::cos(theta_) * dt;
        y_     += linear * std::sin(theta_) * dt;
        theta_ += angular * dt;

        geometry_msgs::Quaternion q =
            tf::createQuaternionMsgFromYaw(theta_);

        // TF
        geometry_msgs::TransformStamped tf_msg;
        tf_msg.header.stamp    = now;
        tf_msg.header.frame_id = "odom";
        tf_msg.child_frame_id  = "base_link";
        tf_msg.transform.translation.x = x_;
        tf_msg.transform.translation.y = y_;
        tf_msg.transform.translation.z = 0.0;
        tf_msg.transform.rotation = q;
        tf_broadcaster_.sendTransform(tf_msg);

        // Odometry message
        nav_msgs::Odometry odom;
        odom.header.stamp    = now;
        odom.header.frame_id = "odom";
        odom.child_frame_id  = "base_link";

        odom.pose.pose.position.x  = x_;
        odom.pose.pose.position.y  = y_;
        odom.pose.pose.orientation = q;

        odom.twist.twist.linear.x  = linear;
        odom.twist.twist.angular.z = angular;

        // Covariance
        odom.pose.covariance  = {0.01, 0, 0, 0, 0, 0,
                                  0, 0.01, 0, 0, 0, 0,
                                  0, 0, 999, 0, 0, 0,
                                  0, 0, 0, 999, 0, 0,
                                  0, 0, 0, 0, 999, 0,
                                  0, 0, 0, 0, 0, 0.05};
        odom.twist.covariance = odom.pose.covariance;

        odom_pub_.publish(odom);
    }

    // ── Publish motor RPMs and currents ────────────────────────────────────────
    void publishMotorFeedback()
    {
        std::lock_guard<std::mutex> lock(feedback_mutex_);

        std_msgs::Float32MultiArray rpm_msg, cur_msg;
        for (size_t i = 0; i < 4; i++) {
            rpm_msg.data.push_back(static_cast<float>(motor_rpms_[i]));
            cur_msg.data.push_back(motor_currents_[i]);
        }
        rpm_pub_.publish(rpm_msg);
        current_pub_.publish(cur_msg);
    }

    // ── Members ───────────────────────────────────────────────────────────────
    ros::NodeHandle nh_, pnh_;
    ros::Subscriber cmd_vel_sub_;
    ros::Publisher  odom_pub_, rpm_pub_, current_pub_;
    tf::TransformBroadcaster tf_broadcaster_;

    serial::Serial serial_;

    // Parameters
    std::vector<int>         wheel_ids_;
    std::vector<std::string> wheel_directions_;
    double wheel_radius_, wheel_track_, wheel_base_;
    double feedback_rate_hz_, cmd_vel_timeout_;
    int    inter_cmd_gap_ms_;

    // Command state
    std::mutex              cmd_mutex_;
    std::condition_variable cmd_cv_;
    double   cmd_linear_   = 0.0;
    double   cmd_angular_  = 0.0;
    bool     cmd_updated_  = false;
    ros::Time last_cmd_time_;

    // Feedback state
    std::mutex              feedback_mutex_;
    std::array<int16_t, 4>  motor_rpms_;
    std::array<float, 4>    motor_currents_;

    // Odometry state
    double    x_ = 0, y_ = 0, theta_ = 0;
    ros::Time last_odom_time_;

    // Threads
    std::atomic<bool>  running_;
    std::thread        tx_thread_;
    std::thread        feedback_thread_;
    std::mutex         bus_mutex_;
};

int main(int argc, char** argv)
{
    ros::init(argc, argv, "base_controller");
    BaseControllerNode node;
    node.spin();
    return 0;
}
